src/postprocess/chi0.jl

```diff
@@ -103,33 +103,101 @@ LinearAlgebra.ldiv!(y::T, P::FunctionPreconditioner, x) where {T} = P.preconditi
+# Solves (1-P) (H-εn) (1-P) δψn = - (1-P) rhs
+# where 1-P is the projector on the orthogonal of ψk
+function sternheimer_solver(Hk, ψk, ψnk, εnk, rhs;
+                            ψk_extra=zeros(size(ψk,1), 0),
+                            εk_extra=zeros(0), schur=true,
+                            tol_cg=1e-12, verbose=false)
     basis = Hk.basis
     kpoint = Hk.kpoint
+    model = basis.model
+    temperature = model.temperature
+
+    # We use a Schur decomposition of the orthogonal of the occupied states
+    # where we still know some information from the partially converged
+    # nonoccupied states (in particular, they are Rayleigh-Ritz wrt to Hk).
+    #
+    # /!\ This is only implemented for insulators at the moment, WIP for
+    # metals, in which case we use empty arrays and all computations are
+    # equivalent to invert the Sternheimer system in Ran(1-P) without the Schur
+    # trick.
+
+    # TODO
+    # finite temperature not supported yet, so we remove extra bands and perform
+    # all the computations with empty arrays
+    if !iszero(temperature) || (schur && (isempty(ψk_extra) || isempty(εk_extra)))
+        ψk_extra = zeros(size(ψk,1), 0)
+        εk_extra = zeros(0)
+        schur = false
+    end
 
-    # we err on the side of caution here by applying Q *a lot*
-    # there are optimizations to be made here
-    Q(ϕ) = ϕ - ψk * (ψk' * ϕ)
-    function QHQ(ϕ)
-        Qϕ = Q(ϕ)
-        Q(Hk * Qϕ - εnk * Qϕ)
+    # Projectors:
+    # projector onto the computed and converged states, which are the occupied
+    # states in the case of insulators
+    P(ϕ) = ψk * (ψk' * ϕ)
+    # projector onto the computed but nonconverged states
+    P_free(ϕ) = ψk_extra * (ψk_extra' * ϕ)
+    # projector onto the computed (converged and unconverged) states
+    P_computed(ϕ) = P(ϕ) + P_free(ϕ)
+    # Q = 1-P is the projector onto the orthogonal of converged states
+    Q(ϕ) = ϕ - P(ϕ)
+    # R = 1-P_computed is the projector onto the orthogonal of computed states
+    R(ϕ) = ϕ - P_computed(ϕ)
+
+    # We put things into the form
+    # δψkn = ψk_extra * αkn + δψkn_uncomputed ∈ Ran(Q)
+    # where δψkn_uncomputed ∈ Ran(R).
+    # Note that, if ψk_extra = [], then 1-P = 1-P_computed and
+    # δψkn = δψkn_uncomputed is obtained by inverting the full Sternheimer
+    # equations in Ran(Q) = Ran(R)
+    #
+    # This can be summarized as the following:
+    #
+    # <---- P ----><------------ Q = 1-P -----------------
+    #              <-- P_free --->
+    # <--------P_computed -------><-- R = 1-P_computed ---
+    # |-----------|--------------|------------------------
+    # 1     N_occupied  N_occupied + N_extra
+
+
+    # ψk_extra are not converged but have been Rayleigh-Ritzed (they are NOT
+    # eigenvectors of H) so H(ψk_extra) = ψk_extra'(Hk-εn)ψk_extra should be a
+    # real diagonal matrix.
+    H(ϕ) = Hk * ϕ - εnk * ϕ
+    ψk_exHψk_ex = real.(Diagonal(εk_extra .- εnk))
```

Comment:
Aargh these names are not so great. But ok, let's bikeshed that later.

Reply:
Also:
```suggestion
    ψk_exHψk_ex = Diagonal(real.(εk_extra .- εnk))
```

Reply:
> Aargh these names are not so great. But ok, let's bikeshed that later.
Yes, I am not very imaginative when it comes to this...